### DISCUSSION TITLE
Number of securities/assets

### DISCUSSION THREAD

**Discussion by christopher-wittlinger** [skfolio/skfolio#89] [General]
How can I add a constraint to the optimization such that the result consists of a predefinded number of assets, e.g. I want a portfolio with 25 positions.

**Comment by HugoDelatte** [comment]
The solution is [here](https://github.com/skfolio/skfolio/discussions/73#discussioncomment-10140135).
It will be added in a future release. For the moment you can use the `add_constraints` parameter.

**Reply by christopher-wittlinger** [reply]
@HugoDelatte Thanks for the response.
Could you also clarify how I can set a required rebalancing scheme? E.g. Every Quarter?

**Reply by christopher-wittlinger** [reply]
Disregard my second question - as you have answered it already in another thread.

**Reply by HugoDelatte** [reply]
You can do it with Walkforward. We recently released time specific rebalancing with the [freq param](https://skfolio.org/generated/skfolio.model_selection.WalkForward.html#skfolio.model_selection.WalkForward).
You can find examples in the docstrings.

For example:
* Rebalancing : Quarterly on the first day
* Test Duration : 1 quarter
* Train Duration : 6 months

```python
cv = WalkForward(test_size=1, train_size=2, freq="QS")
```

**Reply by christopher-wittlinger** [reply]
@HugoDelatte Thanks. One more question:

If I have the following constraints function:

```py
def constraints(w):
    constraints = [
        cp.sum(w) <= 20,
    ]
    return constraints
```

Does this mean that the number of assets in my portfolio should be less than equal to 20? The solver (even MOSEK) always fails with this, even though my sample universe only consists of 20 assets?

**Reply by Conway-Yu** [reply]
No, your constraint specifies that the weight of your assets should be less or equal to 20.

So you could have a portfolio with only one asset with a weight of 20.

**Reply by christopher-wittlinger** [reply]
`cp.length(w) <= 20` does not work as it fails that it is not a DCP problem. Any thoughts?

**Reply by HugoDelatte** [reply]
CVXPY doesn't have a length function. You can reformulate the cardinality constraint as explained in this example: https://github.com/skfolio/skfolio/discussions/73#discussioncomment-10140135

**Reply by christopher-wittlinger** [reply]
I get it now, thanks. However I get now the error `SolverError: Solver 'MOSEK' failed for . Try another solver, or solve with solver_params=dict(verbose=True) for more information` Any way I can check what I can do to improve?

**Reply by HugoDelatte** [reply]
Solvers have difficulties with MIP problems (e.g. cardinality constraints). As the message tells you, you can set `solver_params=dict(verbose=True)` to look at your solver logs (Mosek in this case). 
You can also set `raise_on_failure=False` and `save_problem=True` then access the saved CVXPY problem with `model.problem_` to investigate.

Sometimes you need to play with the scaling params or you solver params.

Last resort is to change your problem and use L1 norm penalization. This will introduce sparsity in your portfolio and reduce your number of assets. You won't get exactly K assets, by the added benefit is that you can find your penalization coef by cross-validation which tend to make your portfolio more stable: https://skfolio.org/auto_examples/1_mean_risk/plot_8_regularization.html#sphx-glr-auto-examples-1-mean-risk-plot-8-regularization-py

**Reply by christopher-wittlinger** [reply]
Thanks for the reply. I played around with the L1 norm earlier and the results were somewhat okayish. The problem I am facing is, that for backtesting/optimization purposes our analysts would like to build constraints around portfolio size (if the universe is approx 600 instruments, no way they want a portfolio with all 600 instruments in it in the worse case). Is the way to go about this to run permutations based on the universe and see which permutation does it best? (even though this seems somewhat foolish?)

**Reply by HugoDelatte** [reply]
Maybe try with the Solver SCIP (pip install cvxpy[SCIP]).

Another approach when you have a big universe is pre-selection: https://skfolio.org/auto_examples/7_pre_selection/plot_2_select_best_performers.html#sphx-glr-auto-examples-7-pre-selection-plot-2-select-best-performers-py

Some papers claim that best performer pre-selection + Minimum Variance outperform most basic models out-of-sample.

Cardinality constraint will work better on smaller universe, so let's say you start with 600 assets: you could do a pre-selection to keep 200 assets, then run your optimization with cardinality constraint of 20 assets on that reduced universe (or a L1 regularization).

**Reply by christopher-wittlinger** [reply]
At the moment I am trying on a reduced set of 9 instruments for testing - later it will be 600. I tried with MOSEK and SCIP, but say it is not feasible. Also when I am using the test data set from s&p from you.

**Reply by CarloNicolini** [reply]
Adding a cardinality constraints often requires using MILP solvers and make the solution (if possible) very slow to find.
I'd suggest trying with an L1 regularization or other methods to induce sparsity.
Take a look at this very [good book by Daniel Palomar](https://bookdown.org/palomar/portfoliooptimizationbook/13.2-sparse-regression.html#methods-for-sparse-regression)

**Reply by christopher-wittlinger** [reply]
Thanks @CarloNicolini - I will take a look.

One question though - If those cardinality constraints are very slow, how are other commercial portfolio optimizers doing it? Usually you can define how many positions you want to have - or is it a different approach than this one? Any insights? (I saw one with a universe of 500 instruments, and it output the backtested, optimized portfolio in like 1-2 minutes.

**Reply by CarloNicolini** [reply]
> Thanks @CarloNicolini - I will take a look.
> 
> One question though - If those cardinality constraints are very slow, how are other commercial portfolio optimizers doing it? Usually you can define how many positions you want to have - or is it a different approach than this one? Any insights? (I saw one with a universe of 500 instruments, and it output the backtested, optimized portfolio in like 1-2 minutes.

That's the secret sauce behind optimizers like Gurobi. And that's why you pay for it. If you have academic account you could subscribe and use it. During my phd it solved problems which were never converging with other optimizers.

**Reply by christopher-wittlinger** [reply]
Thanks @CarloNicolini I will take a look

**Reply by christopher-wittlinger** [reply]
I reduced my test to only have this cardinality constraint. I tried with a lot of solver, such GUROBI, MOSEK, SCIP. 
Maybe I am doing something else wrong?

```py
from sklearn.model_selection import train_test_split

from skfolio import RiskMeasure
from skfolio.datasets import load_sp500_dataset
from skfolio.optimization import MeanRisk, ObjectiveFunction
from skfolio.preprocessing import prices_to_returns
import cvxpy as cp

card = 9

def card_constraints(w):
    card_var = cp.Variable(20, boolean=True)
    constraints = [
        w - card_var <= 0,
        cp.sum(card_var) <= card,
    ]
    return constraints

model = MeanRisk(
    risk_measure=RiskMeasure.VARIANCE,
    objective_function=ObjectiveFunction.MAXIMIZE_RATIO,
    solver=cp.GUROBI,
    add_constraints=card_constraints,
)

prices = load_sp500_dataset()
X = prices_to_returns(prices, drop_inceptions_nan=False)
X_train, X_test = train_test_split(X, test_size=0.33, shuffle=False)
model.fit(X_train)
pred = model.predict(X_test)
pred.plot_composition()
```

**Reply by HugoDelatte** [reply]
I see, you are maximizing a **ratio** and my example was for minimizing a risk measure.
For it to work, we need to add the auxiliary variable (named `factor` in the code), however the parameter `add_constraints` currently doesn't allow you to pass it. I'll create a PR to fix it.
(For the moment, you will have to change `ObjectiveFunction.MAXIMIZE_RATIO` to `ObjectiveFunction.MINIMIZE_RISK`)

**Reply by christopher-wittlinger** [reply]
Thanks @HugoDelatte I will try this.

**Reply by christopher-wittlinger** [reply]
@HugoDelatte That works!

**Reply by christopher-wittlinger** [reply]
Works also for my own universe except if I add more than approx. 400 stocks, then it starts to fail, even with GUROBI. But thanks @HugoDelatte, @Conway-Yu, @CarloNicolini for all the help.

**Reply by christopher-wittlinger** [reply]
One more question: From your experience. If I want to have a portfolio with a low volatility and I want to have a huge universe, should I create a pre selection, that filters out stocks with high volatility and then only have a small subset, and let skfolio only optimize the weights?

**Reply by HugoDelatte** [reply]
Based on experience, if your number of assets is large relative to your number of observations, it will lead to overfitting.

This means that your portfolio's volatility will tend to be very low on the training set but significantly higher on the test set. 

The same rule applies to general machine learning. 

Therefore, any regularization technique, such as adding constraints, penalization, pre-selection, or introducing uncertainty sets, will help. 

To answer your question, yes, what you describe will have a regularization effect and help reduce overfitting. 

One of the goals of the library is to help you analyze such overfitting and determine which regularization techniques work best, using all the scikit-learn tools (cross-validation, grid search, etc.).

**Reply by christopher-wittlinger** [reply]
Cool - that helps a lot. 9 years since I worked with these tools in University, so it really helps to get this insight. Appreciate it!

**Comment by HugoDelatte** [comment]
@christopher-wittlinger  v0.6.0 now contains cardinality , group cardinality and threshold constraints and is available for ratio too (ex: Sharpe Ratio). Special calibration has been implemented to improve convergence. Let me know if that helps for your use-case.
You can use these two examples:

https://skfolio.org/auto_examples/1_mean_risk/plot_15_mip_cardinality_constraints.html#sphx-glr-auto-examples-1-mean-risk-plot-15-mip-cardinality-constraints-py

https://skfolio.org/auto_examples/1_mean_risk/plot_16_mip_threshold_constraints.html#sphx-glr-auto-examples-1-mean-risk-plot-16-mip-threshold-constraints-py

**Reply by christopher-wittlinger** [reply]
@HugoDelatte Many thanks! I will try it out in the upcoming week.
